README
---------------------------------------------------------------
| -- START OF FILE -- -- START OF FILE -- -- START OF FILE -- |
---------------------------------------------------------------
---------------------------------------------------------------
StreamCheck 
FileList:
-WSv2.5.py or WSv3.py
-WS.sh
Verison 0.8.3.1
Built in Python
Created by David Tran (unsignedzero)

64CR(64 character per line plus carriage return)

Please READ THIS TEXT FILE BEFORE ASKING QUESTIONS! If this 
file is incomplete, please go to, 
https://bitbucket.org/unsignedzero/ws/overview
and download the file again, or read it there. 

Abstract:
The purpose of this code is to provide a set of functions for
formatting, debugging and related purposes for programmers.
Each function has specific requirements that can be different
from other functions, so read about it before using it. As 
with other Python Scripts, this can be executed directly.
In this case it will run the formatter function.
---------------------------------------------------------------
---------------------------------------------------------------

 - Table of Contents
 - - Use Ctrl/Cmd + F to jump to certain part of this text
 - - using the characters between the [ and ].

 -  (0*) Information
 - - 00 [Adin] Additional Information
 - - 01 [DEBG] Debug Information

 -  (A*) Functions 
 - - A0 [PASE] pause
 - - A1 [FMAT] formatter
 - - A2 [LCUT] lCount
 - - A3 [RiSK] rSpace_siller

 -  (E*) Exceptions
 - - E0 [PASE] ZText_Error
 - - E1 [FMAT] UnbalancedBraces

 -  (S*) Script
 - - S0 [MAIN] __main__

 -  (Z*) End Information
 - - Z0 [ABUT] About 

---------------------------------------------------------------
---------------------------------------------------------------
 -  (0*) Information
---------------------------------------------------------------
 - - 00 [Adin] Additional Information

There are two versions of this Python Script. Remember to use 
the right version.

As with other Python Scripts, this can be imported but also ran
directly from the script. When executed directly, this script 
will use the formatter function. 

If arguments are passed, the first argument will be the input, 
the second argument will be the output file and the third will
be a bool flag array, in the form of an integer. If the third 
argument is not passed, no special formatting will be used.
If the second argument is not pass, then the output will be 
the input file name with "_edit.txt" concatenated on the end
of it. 

If no arguments as passed, then a prompt will run in 
the script, asking for the above info. 

For more information on the flag parameters see the actual 
function formatter or the docstring of formatter.

---------------------------------------------------------------
 - - 01 [DEBG] Debug Information

Certain functions within this module have debugging functions.

EXCEPTION is a keyword that can be passed to TURN OFF PRESET 
exceptions within that function. (Note: Exceptions due to bad 
input will still occur!)

debug is a keyword that can be passed to on debug mode, for
the function. (Currently this does nothing.)

---------------------------------------------------------------
 -  (A*) Functions 
---------------------------------------------------------------
- - A0 [PASE] pause ()

Function:
WS.pause
()

This function mimics ZX::ZOS(2) without relying on the OS. 
This function pauses the executing module and waits for the 
user to hit the enter key. Anything, entered is discarded.

-PreCondition:
None.

-PostCondition:
The program is pause once before execution resumes.

---------------------------------------------------------------
 - - A1 [FMAT] formatter (String, ...)

Function:
WS.formatter
( fname, fout = None, space_count = 2, *kargs, special = 0, 
  EXCEPTION = False )

This is the core of WS. Minimally, fname (string) the 
name of the input file must be passed. Additional formatting
and information can be passed as well. 

This function assumes that the code passed is has mostly* 
proper syntax and is curly brace balanced! This function 
ONLY works on languages that use curly braces. Languages 
such as Python will NOT format properly because there is
no way to tell if a line is in the right scope of the
right conditional expression.

In addition, this code DOES check for balanced curly 
braces. Should you be passing a file that might not be,
set EXCEPTION to be false and special & 1 to be true, 
for your own debugging purposes

fout -- This the name of the output file. By default, if 
None is passed, this will be set to the fname + "_edit.txt".

space_count -- This is the number of space/tab used for an 
indent. Values should be above 0.

special -- This is a formatting flag for the formatter 
function. Treat this as a bool array that is stored in an 
integer. The result integer passed will be bitwise anded
with the numbers below to turn on/off additional 
functionality. By default this is 0 (Nothing extra).

  1 -- For all closing braces, '}' the condition of the 
    matching opening brace will be shown to the right of the
    closing brace that the formatter is currently scanning, 
    but commented so as not to effect the code. 
    
    Useful for debugging purposes and trying to pair up
    the right curly braces.

  2 -- The tab character will replace the space character that
    is used for indenting. Please note that the space_count 
    will tell formatter how many tabs to place.

    For the purposes of the documentation of this function, 
    spaces and tabs can be used interchangeably depending on
    this flag ONLY. That is, if a function talks about 
    indenting two spaces and this flag is set, the function
    will not indent with two tab characters.

  4 -- Block comments will also be intended, treated as 
    opening and closing braces.

  8 -- Comment lines are shifted back one, so they stand out.
    Will only work if the comment has spaces/tabs to the left.

 16 -- Non-braced if/else/for/while will indent
    This is currently being developed and this flag is in beta

EXCEPTION -- This flag disable balanced curly brace 
exceptions in this function. ALL OTHER EXCEPTIONS will still 
occur! Use carefully of test code with 1 to see which 
braces match with which conditional expression.

-PreCondition:
A valid source code file, that has mostly* correct source 
code. Additional, information can be passed.

-PostCondition:
A properly "formatted" source code file.

!WARNING:Sending invalid or other non-source code files 
will most likely result in erroneous formatting and even
exceptions. In the rare case that it should work out, the
file is curly brace balanced!

*I say mostly because most typos, programmer make will have 
little to no effect on formatter. Formatter checks mainly
for braces and the keywords if/else/for/while and unless 
these are place incorrectly or misspelled, formmater should
have no problems. That being said your mileage will vary.

---------------------------------------------------------------
 - - A2 [LCUT] lcount (String, ...)

Function:
WS.lcount
( fname , fout = None, width = 6, *kargs, code = "UTF-8" )

This function simply takes the input file and writes it to 
the output file, in addition to the line numbers. Line count
starts at 1.

fname -- This is the name of the input file. Unlike formatter,
this function will run regardless of the what the file is. 
However, in the case that the file passed in not a source code,
the length of the line CAN very drastically.

fout -- This is the name of the output file. By default, if 
None is passed then name will be fname + '_counted.txt'

width -- This is the number of additional chars that will be 
added to the beginning of each line for the line count. This 
space will be used for placing the line number. By default 
this is set to 6. 

code -- This sets the character set which will
determine how this function will interpret the input file.
By default this is UTF-8. 

-PreCondition:
A valid file.

-PostCondition:
The output file with line numbers.

---------------------------------------------------------------
 - - A3 [RiSK] rspace_killer (String, ...)

Function:
WS.rspace_killer
( fname, fout = None )

This function removes the pesky whitespace at the end of each
line. While this may seem trivial, the white space does add up.

fname -- The input source code file. If this is NOT a code file,
unexpected results may be created.

fout -- The name of the output file. By default, this function 
will return fname + "_wk.txt" 

-PreCondition:
A valid source code file is passed.

-PostCondition:
The whitespace at the end of each line is removed.
---------------------------------------------------------------
 -  (E*) Exceptions
---------------------------------------------------------------
 - - E0 [PASE] ZText_Error ( Exception )

Exception:
class WS.ZText_Error
( Exception )

This is the base exception class for this module. This 
exception itself is never called but is inherited by other 
exceptions.

---------------------------------------------------------------
 - - E1 [FMAT] UnbalancedBraces ( ZText_Error )

Exception:
class WS.UnbalancedBraces
( ZText_Error )

This is the UnbalancedBrace Exception that is called in
formatter when there is an Unbalanced Curly Brace in the 
source code that formatter is analyzing. 

---------------------------------------------------------------
 -  (S*) Script
---------------------------------------------------------------
 - - S0 [MAIN] __main__

Script Execution:
WS.__main__

This is the script of WS that is executed when this code is 
ran directly, not imported.

This code itself will take in inputs but also created a mini
prompt if none are given. This script itself will run the 
formatter function in WS with standard input , output and  
flags. For more specialized purposes, edit the source code!

Regardless of how the arguments are acquired, via prompt, or
via the shell executing WS, the input is parsed the same way.

If there arguments as passed, then the first argument is the 
input file, the second argument is the output file and the
third is the argument flags for formatter.

if the third argument is omitted then the all the flags will 
be set to 0. If the third AND second argument are omitted, 
then formatter will write to the default output, which is the
input file name with "_edited.txt" concatenated at the end.

Should all there arguments be missing then the mini prompt 
is executed excepting at least one input. The one to three 
inputs given will be treated and parsed the same way, as if 
they were passed in from the shell.

---------------------------------------------------------------

---------------------------------------------------------------
 -  (Z*) End Information
---------------------------------------------------------------
 - - Z0 [ABUT] About 

This program was originally created by unsignedzero to format 
garbage C++ code. Now it can format any code language so long 
as it uses curly braces.

Original Project Name:WonderSwan
Started:July 04, 2011 
Created by unsignedzero
Built in Python V3.2 (NO IDE) Unicode

For more information see 
https://bitbucket.org/unsignedzero/ws/overview
You may contact me via email (unsignedzero@gmail.com) if you
have any suggestions and comments. I DON'T check it daily.

---------------------------------------------------------------
---------------------------------------------------------------
|  -- END OF FILE --   -- END OF FILE --   -- END OF FILE --  |
---------------------------------------------------------------



history
This program was originally created to format C++ files
into Python-like formatting.

Version 0.5.0.0 07-24-2011 Foundation
Original Work Code Created.
All functions in separate files

Version 0.8.0.0 10-09-2011 Raiser
All code now in one file.
Exceptions added and additional modes added.
Extra code lines removed!
No module docstrings have been made YET but all functions have docstrings

Version 0.8.0.1 10-18-2011 
Fixed funky format on GIT

Version 0.8.1.0 10-18-2011
Bugs fixed.
Two versions of code, one for v2 python and one for v3 python.
Codes both work and annoying split bug is also fixed.

Version 0.8.2.0 10-25-2011
Fixed bad input to the script
Included .sh script to run the code manually
Calling the script will automatically run if there are 1 or 2 arguments passed.
Added new mode (16) to formatter
All lines in code are 80 characters or less

Version 0.8.2.1 10-31-2011
Added additional comments to code.
*Will update 16 and add debug soon.
Grammar fixed on all files.

Version 0.8.3.0 11-01-2011
*The updated code from 0.8.1.0 will work from v2.5 and onward, not v2.
(Due to the use of the ternary operator.)
README created.
Additional comments placed in the code.
Default value for the output file can be set by passing 0.
*Will update 16 and add debug soon.

Version 0.8.3.1 11-29-2011
Edited comments and readme for grammatical errors and miscellaneous 
mistakes.
EXCEPTION in v3 is default true, like v2.5
Debug flag created.
*Will update 16 and add debug soon.



WS.py
# WonderSwan
# Provides basic meta programing support
# to python "format"
# 07-24-2011
# Version 0.5
# 10-08-2011
# Version 0.8
# Python Version 3+

#DO NOT CHANGE BELOW THIS POINT

class ZText_Error(Exception):
  """Base class for exceptions in this module"""
  pass

class UnbalancedBraces(ZText_Error):
  """Exception for unbalanced braces
  Attributes:
    expr -- input expression in which the error occurred
    msg  -- explanation of the error
  """
  def __init__(self, expr, msg):
    self.expr = expr
    self.msg = msg

def pause():
  """Correct pauses regardless of the os this function is running on."""
  import os,sys
  pau = "pause" if sys.platform[:3]=="win" else "read -sn 1 -p \"Press any key to continue...\\n\""
  os.system(pau)
  
#def formatter( fname, fout = None, space_count = 2, special = 0, NO_EXCEPTION = False, *kargs):
def formatter( fname, fout = None, space_count = 2, *kargs, special = 0, NO_EXCEPTION = False ):
  r"""
  formatter(...)
     formatter( fname, fout = None, space_co unt = 2, *kargs, special = 0, NO_EXCEPTION = False )

     Given a correct filename fname, this program auto-formats the program 
     file. This function formats source code, in a similar fashion to    
     Python, in which "proper" matching spacing is applied for each line 
     between an opening and closing brace. 

     fname is the ONLY required argument. If fout is not defined, the
     output will have the same name as the input file except now with 
     "_edit.txt" apprended to it.

     ASUMPTIONS:
       The file passed should have balanced braces. If this requirement is   
       not met, the program will return an Exception, unless NO_EXCEPTIONS
       is true!

       All lines of the source code can and will be "reorder" even if they
       are comments!

     ATTRIBUTES:
       fname -- This is the name of input file.

       fout  -- This is the name of the output file. If not specificed,
         then it will be fname + "_edit.txt"

       space_count -- This is the amount of spaces, that each opening brace
         will shift the lines below it.

       special -- special arguments, see below

       NO_EXCEPTION -- Disables exceptions messages of unbalanced braces
 
      SPECIAL:
        Treat this variable as an array of bools.  (Represented as an integer)
        This turns on/off additional functions, listed below.
        
        1 -- Comments after the end brace, what the opening brace was
        
        2 -- Use tabs, rather than spaces!

        4 -- Treat /* */ comments as braces, for the purposes of     
               indentation

        8 -- Comment-only // lines are shifted -1 left
        
        
  """
  import sys
  
  shift       = 0
  shift_delay = 0
  brace_start = '{'
  brace_end   = '}'
  stack       = []

  #Files 
  source_code = open(fname, "r" )
  fout = (fname + "_edit.txt") if (fout == None) else fout 
  dest_code   = open(fout, "w" )
  ###err_code    = open(fname + "_err.txt", "w" )

  print("%s starting with %s. Output is %s." % (sys._getframe(0).f_code.co_name , fname, fout) )

  for (count,line) in enumerate(source_code) :

  ###err_code.write( '%03d | ' % len(line.strip() ) + line)

    #Empty Line are Empty
     empty_line = 1 if line.strip() else 0
 
     line = ( ( empty_line * shift * space_count * ' ' ) +
                line.strip()                             )
 
		     #Insert Extra Formatting here
     if special > 0:
       if special & 1 :
         if '{' in line and '}' not in line :
           stack.append( line[:-1].strip() )
         elif '{' not in line and '}' in line :
           line += " // " + stack.pop()
       if special & 2 :
           line = ( '\t' * shift ) + line.lstrip()
       if special & 4 :
         if '\*' in line:
           shift_delay +=1
         if '*\\' in line:
           shift_delay -=1
       if special & 8 :
         if (line.lstrip()).startswith('//'):
           if (line[0] == ' '):
             line = line[1:]
         
     line += '\n'

    #Write to File
     dest_code.write( line )

    ##Calculate Shift for next line
     if brace_start in line :
						        shift += 1
     if brace_end   in line :
						        shift -= 1
     if shift_delay != 0    :
       shift += shift_delay
       shift_delay = 0
       
     if NO_EXCEPTION and shift < 0 :
       print( "\n  File \"%s\", line %i, in %s" % ( fname, count,  sys._getframe().f_code.co_name ) )
       raise UnbalancedBraces( 0 , "Unbalanced Closing Braces in the file" )
  if NO_EXCEPTION and shift != 0:
    print( "\n  File \"%s\" , in %s" % ( fname,  sys._getframe().f_code.co_name ) )
    raise UnbalancedBraces( 0 , "Unbalanced Opening Braces in the file!" )
  print( "%s Compeleted!" % sys._getframe(0).f_code.co_name )

#def lcount( fname , fout = None, width = 6, code = "UTF-8", *kargs ):
def lcount( fname , fout = None, width = 6, *kargs, code = "UTF-8" ) :
  r"""
  lcount(...)
     lcount( fname , fout = None, width = 6, *kargs, code = "UTF-8" )

     Writes the line number of each line into the output text file.

     ATTRIBUTES:
       fname -- This is the name of input file.
       
       fout  -- This is the name of the output file. If not specificed,
         then it will be fname + "_counted.txt"
         
       width -- Sets the width of the number column
       
       code  -- Sets the default coding of the file

  """

  import sys

  #Files
  file_in  = open(fname, "r", 1, code)
  fout = (fname + '_counted.txt') if (fout == None) else fout 
  file_out = open(fout,"w" , 1, code)

  print("%s starting with %s. Output is %s." % (sys._getframe(0).f_code.co_name , fname, fout) )
    
  width = "%0" + str(width) + "d | "

  for (count,line) in enumerate(file_in) :
    file_out.write( str( width % count) + line )  

  print( "%s Compeleted!" % sys._getframe(0).f_code.co_name )  

def rspace_killer ( fname, fout = None ) :
  r"""
  rspace_killer(...)
     rspace_killer ( fname, fout = None )
     
     Removes excess white space on the right
     
     ATTRIBUTES:
 
     fname -- This is the name of the input file.
     
     fout  -- This is the name of the output file. If not specificed,
       then it will be fname + "_wk.txt"     
  """

  import sys
  
  fin = open(source,"r")
  fout = source + '_wk.txt' if ( fout == None ) else fout
  dest = open(fout,"w")

  print("%s starting with %s. Output is %s." % (sys._getframe(0).f_code.co_name , fname, fout) )
    
  for line in fin :
    fout.write( line.rstrip() )
    
  print( "%s Compeleted!" % sys._getframe(0).f_code.co_name )  
  
if __name__ == "__main__" :
  import sys
  print("Starting WS")
  try:
    inf = sys.argv[1]
    outf = sys.argv[2]
    spf = sys.argv[3]
  except IndexError:
    #finput = raw_input("Please enter a file name\n")
    finput = input("Please enter a file name\n")
    fname = finput.split()
    try:
      special = finput[1]
    except IndexError:
      special = 0
    finally:
      formatter(finput[0], special)
  else:
    formatter(inf,outf, special = spf)
  pause()
  


WSv2.py
# WonderSwan
# Provides basic meta programing support
# to python "format"
# 07-24-2011
# Version 0.5
# 10-08-2011
# Version 0.8
# 10-18-2011
# Version 0.8.0.1
# 10-18-2011
# Version 0.8.1.0
# 10-25-2011
# Version 0.8.2.0

#DO NOT CHANGE BELOW THIS POINT

class ZText_Error ( Exception ):
  """Base class for exceptions in this module"""
  pass
  
class UnbalancedBraces ( ZText_Error ):
  """Exception for unbalanced braces
  Attributes:
    expr -- input expression in which the error occurred
    msg  -- explanation of the error
  """
  def __init__(self, expr, msg):
    self.expr = expr
    self.msg = msg

def pause():
  """Correct pauses regardless of the os this function is running on."""
  #import os,sys
  #pau = "pause" if sys.platform[:3]=="win" else 
    #"read -rsn 1 -p \"Press any key to continue...\\n\""
  #os.system(pau)
  raw_input("Press any key to continue...\n")

#def formatter( fname, fout = None, space_count = 2, 
  #*kargs, special = 0, NO_EXCEPTION = False ):
def formatter( fname, fout = None, space_count = 2, 
  special = 0, NO_EXCEPTION = False ):
  r"""
  formatter(...)
     formatter( fname, fout = None, space_co unt = 2, 
       *kargs, special = 0, NO_EXCEPTION = False )

     Given a correct filename fname, this program auto-formats the program 
     file. This function formats source code, in a similar fashion to    
     Python, in which "proper" matching spacing is applied for each line 
     between an opening and closing brace. 

     fname is the ONLY required argument. If fout is not defined, the
     output will have the same name as the input file except now with 
     "_edit.txt" apprended to it.

     ASUMPTIONS:
       The file passed should have balanced braces. If this requirement is   
       not met, the program will return an Exception, unless NO_EXCEPTIONS
       is true!

       All lines of the source code can and will be "reorder" even if they
       are comments!

     ATTRIBUTES:
       fname -- This is the name of input file.

       fout  -- This is the name of the output file. If not specificed,
         then it will be fname + "_edit.txt"

       space_count -- This is the amount of spaces, that each opening brace
         will shift the lines below it.

       special -- special arguments, see below

       NO_EXCEPTION -- Disables exceptions messages of unbalanced braces
       
      SPECIAL:
        Treat this variable as an array of bools.  (Represented as an integer)
        This turns on/off additional functions, listed below.
        
         1 -- Comments after the end brace, what the opening brace was
        
         2 -- Use tabs, rather than spaces!

         4 -- Treat /* */ comments as braces, for the purposes of     
               indentation

         8 -- Comment-only // lines are shifted -1 left

        16 -- Non-braced, if/for/while will indent
        
  """
  import sys
  
  shift       = 0
  shift_delay = 0   #For   4
  cond_shift  = 0   #For  16
  mline_shift = 0   #Future Use
  brace_start = '{'
  brace_end   = '}'
  stack       = []  #For   1
  space_char  = ' ' #For   2

  #Files 
  source_code = open(fname)
  fout = (fname + "_edit.txt") if (fout == None) else fout 
  dest_code   = open(fout, "w" )
  ###err_code    = open(fname + "_err.txt", "w" )

  print("%s starting with %s. \nOutput is %s." % 
    (sys._getframe(0).f_code.co_name , fname, fout) )

  #SPECIAL
  if special & 2 :
    space_char = '\t'

  for (count,line) in enumerate(source_code) :

     ###err_code.write( '%03d | ' % len(line.strip() ) + line)

    #Empty Line are Empty
     empty_line = 1 if line.strip() else 0
   
     line = ( ( empty_line * ( shift + cond_shift + mline_shift  )*  
              space_count * space_char                            ) 
              + line.strip()                                      )
              
    #Insert Extra Formatting here
     if special > 0:
       if special & 1 :
         if '{' in line and '}' not in line :
           stack.append( line[:-1].strip() )
         elif '{' not in line and '}' in line :
           line += " // " + stack.pop()
       if special & 4 :      
         if r'/*' in line:
           shift_delay +=1
         if r'*/' in line:
           shift_delay -=1
       if special & 8 :
         if (line.lstrip()).startswith('//'):
           if (line[0] == ' ' or line[0] == '\t' ): #CHECK ME
             line = line[1:]
       if special & 16 :
         if ( 'if' in line or 'else' in line 
          or 'for' in line or 'else' in line ) and '{' not in line:
           cond_shift = 1
         else:
           cond_shift = 0
    #Write to File
     dest_code.write( line + '\n' )

    ##Calculate Shift for next line
     if brace_start in line :
       shift += 1
     if brace_end   in line :
       shift -= 1
     if shift_delay != 0    :
       shift += shift_delay
       shift_delay = 0
       
     if NO_EXCEPTION and shift < 0 :
       print( "\n  File \"%s\", line %i, in %s" % 
         ( fname, count,  sys._getframe().f_code.co_name ) )
       raise UnbalancedBraces( 0 , "Unbalanced Closing Braces in the file" )
  if NO_EXCEPTION and shift != 0:
    print( "\n  File \"%s\" , in %s" % 
      ( fname,  sys._getframe().f_code.co_name ) )
    raise UnbalancedBraces( 0 , "Unbalanced Opening Braces in the file!" )
  print( "%s Compeleted!" % sys._getframe(0).f_code.co_name )

#def lcount( fname , fout = None, width = 6, *kargs, code = "UTF-8" ) :
def lcount( fname , fout = None, width = 6, code = "UTF-8" ) :
  r"""
  lcount(...)
     lcount( fname , fout = None, width = 6, *kargs, code = "UTF-8" )

     Writes the line number of each line into the output text file.

     ATTRIBUTES:
       fname -- This is the name of input file.
       
       fout  -- This is the name of the output file. If not specificed,
         then it will be fname + "_counted.txt"
         
       width -- Sets the width of the number column
       
       code  -- Sets the default coding of the file

  """

  import sys

  #Files
  file_in  = open(fname, "r", 1, code)
  fout = (fname + '_counted.txt') if (fout == None) else fout 
  file_out = open(fout,"w" , 1, code)

  print("%s starting with %s. Output is %s." % 
    (sys._getframe(0).f_code.co_name , fname, fout) )
    
  width = "%0" + str(width) + "d | "

  for (count,line) in enumerate(file_in) :
    file_out.write( str( width % count) + line )  

  print( "%s Compeleted!" % sys._getframe(0).f_code.co_name )  

def rspace_killer ( fname, fout = None ) :
  r"""
  rspace_killer(...)
     rspace_killer ( fname, fout = None )
     
     Removes excess white space on the right
     
     ATTRIBUTES:
     
     fname -- This is the name of the input file.
     
     fout  -- This is the name of the output file. If not specificed,
       then it will be fname + "_wk.txt"     
  """

  import sys
  
  fin = open(source,"r")
  fout = source + '_wk.txt' if ( fout == None ) else fout
  dest = open(fout,"w")

  print("%s starting with %s. Output is %s." % 
    (sys._getframe(0).f_code.co_name , fname, fout) )
    
  for line in fin :
    fout.write( line.rstrip() )
    
  print( "%s Compeleted!" % sys._getframe(0).f_code.co_name ) 
 
def ver_update():
  import sys
  if ( sys.version_info[0] == 2 ):
    import __future__ 

if __name__ == "__main__" :
  import sys
  #ver_update()
  print("Starting WS")
  try:
    inf = sys.argv[1]
    outf = (sys.argv[1] + "_edit.txt") if ( 
      len(sys.argv[1:2]) == 1 ) else sys.argv[2]
    spf = 0 if ( len(sys.argv[1:3]) < 3 ) else sys.argv[3]
  except IndexError:
    finput = raw_input("Please enter a file name\n")
    finput = finput.split()
    try:
      mode = int(finput[1])
    except IndexError:
      mode = 0
    finally:
      if ( isinstance(finput,list) ) :
        formatter(finput[0], special = int(mode))
      else :
        formatter(finput, special = 0)
  else:
    formatter(inf,outf, special = spf)
  pause()
  
